Update nav styles to match core designs

joshuatf opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Related to https://github.com/woocommerce/navigation/pull/68 - this PR updates to the latest design exploration for the nav component ( https://www.figma.com/file/e4tLacmlPuZV47l7901FEs/WordPress-Design-Library?node-id=1456%3A72 ).

## Screenshots <!-- if applicable -->

<img width="306" alt="Screen Shot 2020-09-02 at 10 09 59 AM" src="https://user-images.githubusercontent.com/10561050/92013638-bbaf0000-ed56-11ea-8a1d-e79ae88791e3.png">

joshuatf committed:
Update nav styles to match new core designs

joshuatf committed:
Add menu title prop and menu title styling

psealock commented:
I think it makes sense to go with the dark theme and write CSS to the spec pasted above. The reason is because both the FSE and WooCommerce usage won't be in the context of the wp-admin sidebar. Should a time come to examine that change, it will likely be in conjunction with a change to the top nav bar.

FSE and WooCommerce will occupy full screens. This means that this Navigation component won't need to match a background theme color found in wp-admin sidebar or top nav bar.

![templates](https://cldup.com/4K_wRjUxRd-3000x3000.png)

![Screen Shot 2020-09-02 at 12 28 06 PM](https://user-images.githubusercontent.com/1922453/91918717-d2823200-ed17-11ea-8998-560d42d9bb86.png)

I do think that focus style colors should respect the theme. But ping @jameskoster to confirm.

![Screen Shot 2020-09-02 at 12 31 43 PM](https://user-images.githubusercontent.com/1922453/91918887-60f6b380-ed18-11ea-8d79-c7bc1dad0994.png)

jameskoster commented:
> I do think that focus style colors should respect the theme.

Agreed!

joshuatf committed:
Add dark styling

joshuatf committed:
Add secondary menu to nav story

joshuatf commented:
Thanks for the input @jameskoster and @psealock!  This has been rebased and ready for another review.

Dark styles are added and currently this just respects the built-in theme colors of the GB buttons.  If there are more theming issues, I suggest we handle in a follow-up.

This PR also adds menu titles (section titles) and a secondary menu to the nav story to illustrate potential use cases.

joshuatf committed:
Add tests for menu titles

psealock commented on packages/components/src/navigation/index.js:
```diff
@@ -73,13 +73,14 @@ const Navigation = ( { activeItemId, children, data, rootTitle } ) => {
 
 		return (
-			<Button
+			<BackButtonUi
```
Other names have a capitalized `I` as in `BackButtonUI`.

joshuatf replied:
Good catch.

psealock reviewed (commented):
This is looking really nice @joshuatf. I just left a minor comment.

>This PR also adds menu titles (section titles) and a secondary menu to the nav story to illustrate potential use cases.

This is the pattern we set for WooCommerce, but I think its not one we should settle on because it falls short in several ways:

* The consuming app needs to have knowledge of any secondary menus in its render function as well as at the point of constructing data.
* This works well for a single secondary menu on the navigation root. What if there is more than one? What if those secondary menu titles need to be dynamic? 

I don't think rendering these secondary menus or groupings should be completely the responsibility of the consumer. I created an issue to track that feature in https://github.com/WordPress/gutenberg/issues/24992 and I think it'd be best to explore that outside of this PR (even if we do go this route in the end).

joshuatf committed:
Fix back button style component name

joshuatf commented:
Thanks for the review @psealock.  Updated the ui component name based on your feedback.

>This works well for a single secondary menu on the navigation root. What if there is more than one? What if those secondary menu titles need to be dynamic?

The consumer can decide if there is more than one.  In the example, I just added `isSecondary` but this could be a slug or any other property to discern various menus.  The title can also be dynamic here- also up to the consumer.

I do think this should be more up to the consumer and prefer a slightly more compositional approach, though that does come with some drawbacks.  We can discuss in https://github.com/WordPress/gutenberg/pull/25035

psealock commented on packages/components/src/navigation/stories/index.js:
```diff
@@ -92,43 +128,55 @@ function Example() {
+										.map( ( item ) =>
+											renderMenuItem( item )
+										) }
+								</NavigationMenu>
```
>The consumer can decide if there is more than one. In the example, I just added isSecondary but this could be a slug or any other property to discern various menus. The title can also be dynamic here- also up to the consumer.

This is true, and this block is a nice example of the consumer having control. What if the entire navigation has 10 of these submenus throughout the hierarchy? Looking at the designs for FSE, its possible that may be the case. You'd need a block like this for every single submenu. `level.children` is getting filtered with each block. I think a better way exists and worth exploring.

psealock reviewed (approved):
Although the example submenu is likely to be removed, the other changes are good to go. Nice job @joshuatf !
